sdk/go/common/resource/plugin/provider-implementers-guide.md

```diff
@@ -0,0 +1,609 @@
+operations necessary to reconcile the two.
+
+Each entry in a detailed diff has a change kind that describes how the value of and
+input property differs, whether or not the difference requires replacement, and which old
+value was used for determining the difference. The core change kinds are:
+
+- `Add`, which denotes an `Object` property or `Array` element that was added
+- `Update`, which denotes an `Object` property or `Array` element that was updated
+- `Delete`, which denotes an `Object` property or `Array` element that was removed
+
+Each of these core kinds is paramaterized on whether or not the change requires
+replacement and whether the old value of the property should was read from the
+resource's old input `Object` or old state `Object`.
+
+*TODO*: the input/output flag is a bit clumsy, as it is the only part of the system
+that implies some correspondence between input and output `Object` schemas. It was
+chosen over an approach that used old/new values due in order to remove the possibility
+of a provider accidentally revealing a secret value as part of a diff. We should
+reconsider this approach if we can find an easy way to maintain secretness.
+
+#### Create
+
+The `Create` method is responsible for creating a new instance of a resource from an
+input `Object` and returning the resource's state `Object`. `Create` may be called during
+a [preview](#preview) in order to compute a hypothetical state `Object` without actually
+creating the resource, in which case the `preview` argument will be `true`.
+
+##### Parameters
+
+- `urn`: the [URN](#urns) of the resource.
+- `news`: the input `Object` for the resource. This value must have been generated by a
+          prior call to `Check`. If `preview` is true, this value may contain
+          [`Unknown`](#unknowns) value; otherwise, it is guaranteed to be fully-known.
+- `timeout`: the timeout for the create operation. If this value is `0`, the provider
+             should apply the default creation timeout for the resource.
+- `preview`: if true, the provider should calculate the state `Object` as accurately as it
+             is able without actually creating the resource. Top-level properties that
+             are present in the resource's [schema](#schema) but are omitted from its
+             state `Object` should be treated as having the value [`Unknown`](#unknowns).
+             Nested properties with values that are not computable must be explicitly set
+             to [`Unknown`](#unknowns). If it is not possible to guarantee that the value
+             produced by a preview will match the value that would be produced by actually
+             creating the resource, the value should be left unknown.
+
+##### Results
+
+- `id`: the ID for the created resource. If `preview` is true, this value will be ignored.
+- `state`: the new state `Object` for the resource. If `preview` is true, this value may
+           contain [`Unknown`s](#unknowns).
```

Comment:
It's not part of the core RPC contract itself - but it seems like in principle all native Pulumi providers will want to use the "partial error" support to ensure they return `id` and `state` along with any error that occurs after a backing physical resource has been allocated - like we do in the Kubernetes provider, and likely should ~anywhere else to avoid leaking resources while waiting for them to initialize (and potentially fail to initialize).  Should this be called out here somehow?